### PR TITLE
Add EnergyDependentMorphologyEstimator

### PR DESCRIPTION
**Pull request by Astro-Kirsty**
Add the `EnergyDependentEstimator` to resolve [#4707](https://github.com/gammapy/gammapy/issues/4707)

This is an updated PR to include all recent upgrade. Old PR [here](https://github.com/gammapy/gammapy/pull/4708)

**Description**
An example is included for MSH 1552 to show how the estimator works: 
https://gist.github.com/Astro-Kirsty/f74ee75ca543bd8f763cb9225d3b84f8

I've added here a notebook showing the validation for the method/results:
https://gist.github.com/Astro-Kirsty/33efa8847e5f756fbe214f7d12f54bfc

Documentation and a recipe for the validation of results need to be added -- but in a separate PR

### CODE REVIEW

**Comment by Astro-Kirsty** [comment]
Did not include an example a general fit to determine the overall parameters for the model (see [here](https://github.com/gammapy/gammapy/pull/4708#discussion_r1334007550)) as I believe the user should already have a handle on doing this before they try to use the estimator. 

[This](https://github.com/gammapy/gammapy/pull/4708#discussion_r1347230550) also need implementation or discussion.

I will add tests soon

**Comment by bkhelifi** [comment]
Once this PR is merged, indeed a tutorial would be needed... or insert this functionality within an existing tutorial

File: gammapy/estimators/energydependence.py
Comment (bkhelifi) [inline gammapy/estimators/energydependence.py:11]: Can you define in the docstring the parameters?

File: gammapy/estimators/energydependence.py
Comment (bkhelifi) [inline gammapy/estimators/energydependence.py:41]: In the previous PR, we were speaking about adding in the docstring the list of available fitting backend for all estimators. Can you put an issue to remind that?

Comment (bkhelifi) [inline gammapy/estimators/energydependence.py:41]: You could precise that when None iminuit is used

File: gammapy/estimators/energydependence.py
Comment (bkhelifi) [inline gammapy/estimators/energydependence.py:78]: here, the initial mask of the dataset is erased. Is it the wished behaviour? or an AND is preferable?

File: gammapy/estimators/energydependence.py
Comment (bkhelifi) [inline gammapy/estimators/energydependence.py:146]: The same remark as above..

**Comment by bkhelifi** [comment]
I am wondering if some visualization utils of the results are needed. Any thoughts?

File: gammapy/estimators/energydependence.py
Comment (registerrier) [reply gammapy/estimators/energydependence.py:41]: I don't think this is a good idea to duplicate the list of available backends in the documentation. This will become impossible to maintain.
What should be done is to put a proper link to `~gammapy.modeling.Fit` in the argument that will directly link to the list of available backends.

**Comment by Astro-Kirsty** [comment]
Thanks @bkhelifi. We have the `SpatialModel` plotting implemented, however, we could also add other visualisation i.e. parameter vs energy plot

**Comment by Astro-Kirsty** [comment]
I've added some tests, however these utilise the data in the example. Not sure if there is another dataset that can be used or if it would be useful to save the MSH 15-52 dataset in gammapy data. 
Suggestions are welcome

**Comment by adonath** [comment]
@Astro-Kirsty Ideally you use a small (simulated) toy dataset with a source that has energy depend morphology to check whether you can approximately reconstruct the true values. Here is in example of how to define such a model: https://docs.gammapy.org/1.1/tutorials/api/models.html#models-with-energy-dependent-morphology

**Comment by Astro-Kirsty** [comment]
Thanks @adonath, I already have a dataset for this so I will adapt the test accordingly

File: gammapy/estimators/energydependence.py
Comment (Astro-Kirsty) [reply gammapy/estimators/energydependence.py:95]: This was the wished behaviour, however, I am not sure whether it is preferable

File: examples/tutorials/analysis-3d/energy_dependence_estimation.py
Comment (MRegeard) [inline examples/tutorials/analysis-3d/energy_dependence_estimation.py:120]: Don't you need to put the energy_axis_true here ?

```suggestion
global_dataset = MapDataset.create(geom, energy_axis_true=energy_axis_true)
```

File: examples/tutorials/analysis-3d/energy_dependence_estimation.py
Comment (MRegeard) [inline examples/tutorials/analysis-3d/energy_dependence_estimation.py:196]: This is not needed, either this one or the one above should be remove.

File: examples/tutorials/analysis-3d/energy_dependence_estimation.py
Comment (MRegeard) [inline examples/tutorials/analysis-3d/energy_dependence_estimation.py:203]: Maybe a little bit of context is needed here. Especially, what is `df`?

File: examples/tutorials/analysis-3d/energy_dependence_estimation.py
Comment (MRegeard) [inline examples/tutorials/analysis-3d/energy_dependence_estimation.py:218]: Why not using `table` here instead of `results["energy_dependence"]["result"]`?

File: examples/tutorials/analysis-3d/energy_dependence_estimation.py
Comment (MRegeard) [inline examples/tutorials/analysis-3d/energy_dependence_estimation.py:211]: If `table` is used below (see next comment), maybe it would be wise to renamed it to a more specific name, i.e. `results_table`.

**Review by MRegeard — COMMENTED** [review]
Thanks @Astro-Kirsty, this is a very nice and useful feature !
See inline comments.

File: gammapy/estimators/energydependence.py
Comment (Astro-Kirsty) [reply gammapy/estimators/energydependence.py:16]: The idea was to return a dict, similar to how it is done in the `~gammapy.modeling.select_nested_models`.
Open to other ideas, because this was a long debated topic. What are the most important values and how to show them!

**Comment by Astro-Kirsty** [comment]
Note: need to wait for #4927 to be merged, so that the test can be fixed. 
This PR will also need a rebase - but I will wait until other reviews are received and we are ready for merge.

**Review by registerrier — COMMENTED** [review]
Thanks @Astro-Kirsty . Here are some further (but old) comments.

The CI is broken because of the angular_separation import with astropy 6

File: gammapy/estimators/tests/test_energydependence.py
Comment (registerrier) [inline gammapy/estimators/tests/test_energydependence.py:5]:
```suggestion
from astropy.coordinates import angular_separation
```

File: gammapy/estimators/energydependence.py
Comment (registerrier) [inline gammapy/estimators/energydependence.py:217]: Is it always working? Shouldn't you rather access `results[i].models[self.source].free_parameters`?

Comment (registerrier) [reply gammapy/estimators/energydependence.py:217]: You could use `Parameters.value` to directly access all parameters values.

File: gammapy/estimators/tests/test_energydependence.py
Comment (registerrier) [inline gammapy/estimators/tests/test_energydependence.py:71]: you could make this a fixture

File: gammapy/estimators/tests/test_energydependence.py
Comment (registerrier) [inline gammapy/estimators/tests/test_energydependence.py:127]: You should not leave test code outside functions or class.

You have two options:
- create a fixture/function to create the model and then in each test function call the `create_model` and `create_dataset` functions
- Create a general Test class that will perform this code on init and then have test methods for the various things required.

**Review by registerrier — COMMENTED** [review]
Thanks @Astro-Kirsty . I have left some inline comments. The main one concerns the test whic should be structured differently. You shouldn't have code executed outside functions or class. You could restructure it to go into a test class.

File: gammapy/estimators/energydependence.py
Comment (Astro-Kirsty) [reply gammapy/estimators/energydependence.py:129]: This will not work because we are working with sliced datasets that do not have the name `source`.
See line 100

File: gammapy/estimators/energydependence.py
Comment (Astro-Kirsty) [reply gammapy/estimators/energydependence.py:132]: We need to know access specifically the `spatial_model`, so it should not be `parameters`

File: gammapy/estimators/energydependence.py
Comment (registerrier) [inline gammapy/estimators/energydependence.py:101]: But what if there are more models set on the dataset?

File: gammapy/estimators/energydependence.py
Comment (Astro-Kirsty) [reply gammapy/estimators/energydependence.py:132]: Also, same as above the names are no longer so we take `sliced.models[0]` which is the `SkyModel` rather than the `FoVBackgroundModel`

File: gammapy/estimators/energydependence.py
Comment (Astro-Kirsty) [reply gammapy/estimators/energydependence.py:198]: Same as above!

File: gammapy/estimators/energydependence.py
Comment (Astro-Kirsty) [reply gammapy/estimators/energydependence.py:217]: Currently it is implemented through the parameters variable on line 193, such that the results will only include the spatial model!

File: examples/tutorials/analysis-3d/energy_dependence_estimation.py
Comment (Astro-Kirsty) [reply examples/tutorials/analysis-3d/energy_dependence_estimation.py:19]: Okay, should I move it somewhere else or remove it?

File: gammapy/estimators/energydependence.py
Comment (Astro-Kirsty) [reply gammapy/estimators/energydependence.py:209]: I cannot seem to get this to work with the naming, I believe because there are multiple parameters looped over and the names would be the same?

File: examples/tutorials/analysis-3d/energy_dependence_estimation.py
Comment (AtreyeeS) [inline examples/tutorials/analysis-3d/energy_dependence_estimation.py:267]: Do you need 3 Tables with one row each? Showing all the parameters in a Table simultaneously will be better

File: gammapy/estimators/energydependence.py
Comment (AtreyeeS) [inline gammapy/estimators/energydependence.py:77]: Why configure `print_level`?

File: gammapy/estimators/energydependence.py
Comment (AtreyeeS) [inline gammapy/estimators/energydependence.py:106]: This will remove the other `models` from the fit. Is that what you want? 
One might want to search for source extension in the presence of other sources (eg, diffuse emission)

File: gammapy/estimators/energydependence.py
Comment (AtreyeeS) [inline gammapy/estimators/energydependence.py:101]: How is the `mask_fit` for `sliced_src` handled?

**Review by AtreyeeS — COMMENTED** [review]
Thanks @Astro-Kirsty ! Sorry for the delay, I had forgotten to finish a review I had started long back.

Please see inline comments for some specific questions.

In general, I think it can be useful if you add a dataset with an energy dependent model in `gammapy-data`. A major part of the test is defining a model, doing a simulation etc. Same for the tutorial, where the data reduction takes up a significant part.

File: gammapy/estimators/energydependence.py
Comment (Astro-Kirsty) [reply gammapy/estimators/energydependence.py:15]: The idea was that this is an additional calculation that could be made. 
One should really utilise the results_edep significance values because these include any correlation between parameters.

File: gammapy/estimators/energydependence.py
Comment (Astro-Kirsty) [reply gammapy/estimators/energydependence.py:101]: I believe this comment is outdated? There is no `mask_fit` defined within the code now?

File: gammapy/estimators/energydependence.py
Comment (Astro-Kirsty) [reply gammapy/estimators/energydependence.py]: This comment is possibly outdated?

File: examples/tutorials/analysis-3d/energy_dependence_estimation.py
Comment (Astro-Kirsty) [reply examples/tutorials/analysis-3d/energy_dependence_estimation.py:267]: The way the `weighted_chi2_parameter` function is created means that one table is created for a given parameter. I will adjust the implementation a bit!

**Comment by Astro-Kirsty** [comment]
Thanks for the comments @AtreyeeS.

> In general, I think it can be useful if you add a dataset with an energy dependent model in gammapy-data. 

I agree, it would certainly speed up the tutorial also.

File: gammapy/estimators/energydependence.py
Comment (AtreyeeS) [reply gammapy/estimators/energydependence.py:15]: If there are caveats to keep in mind, it should be clearly mentioned in the doc string and the tutorial, otherwise everyone will definitely use it.

File: gammapy/estimators/energydependence.py
Comment (AtreyeeS) [reply gammapy/estimators/energydependence.py:101]: I was unsure if the slicing takes the masks correctly into account. Inow checked that it does, so the comment is indeed irrelevant

**Comment by AtreyeeS** [comment]
> > In general, I think it can be useful if you add a dataset with an energy dependent model in gammapy-data.
> 
> I agree, it would certainly speed up the tutorial also.

Or, if we don't want to add to gammapy data for one particular use case, you can do the data reduction in the tutorial using a config file to avoid those extra lines of code

**Comment by Astro-Kirsty** [comment]
I decided to add a file to the `gammapy-data` [repo](https://github.com/gammapy/gammapy-data/pull/46), as this is useful in both the tests and the tutorial, significantly decreasing the unnecessary parts of the the tutorial. 
This will need to be merged and then I will push the new commits for this

File: gammapy/estimators/energydependence.py
Comment (Astro-Kirsty) [reply gammapy/estimators/energydependence.py]: No, that is not the desired behaviour. 
I will fix this!

**Review by bkhelifi — CHANGES_REQUESTED** [review]
Thanks @Astro-Kirsty... Except very minor suggestions, I think that this is ready to be merged

**Review by AtreyeeS — DISMISSED** [review]
Thanks @Astro-Kirsty ! No further comments

File: gammapy/estimators/energydependence.py
Comment (Astro-Kirsty) [reply gammapy/estimators/energydependence.py:119]: There is no `Models` attribute, only the `DatasetModels`. Therefore this is not possible, hence the weird way I had to implement it

**Comment by Astro-Kirsty** [comment]
I have implemented all comments here. 
@AtreyeeS, @registerrier, @bkhelifi any final comments?

**Review by AtreyeeS — APPROVED** [review]

**Review by registerrier — APPROVED** [review]
Thanks @Astro-Kirsty . This looks good. No further comments.

**Pull request merged**
